python/dgl/data/molecule.py

```diff
@@ -0,0 +1,471 @@
+
+            with open(path_to_dgl_graphs, 'wb') as f:
+                pickle.dump(graphs, f)
+
+        self.graphs = graphs
+
+        # Prepare labels
+        labels = np.hstack(
+            [np.reshape(np.array(self.df[task].values), (num_datapoints, 1)) for task in self.tasks])
+
+        # Some data points do not have all labels and we want to exclude them for loss computation.
+        w = np.ones((num_datapoints, n_tasks))
+        missing = np.zeros_like(labels).astype(int)
+
+        for i in range(num_datapoints):
+            for task in range(n_tasks):
+                if labels[i, task] == "":
+                    missing[i, task] = 1
+
+        for i in range(num_datapoints):
+            for task in range(n_tasks):
+                if missing[i, task]:
+                    # The missing labels are now replaced with label 0 and will be masked
+                    # for loss computation
+                    labels[i, task] = 0.
+                    w[i, task] = 0.
+
+        self.labels = labels.astype(float)
+        self.w = w.astype(float)
+
+    def weight_balancing(self):
+        """Perform re-balancing for each task.
+
+        It's quite common that the number of positive samples and the
+        number of negative samples are significantly different. To compensate
+        for the class imbalance issue, we can weight each datapoint in
+        loss computation.
+
+        In particular, for each task we will set the weight of negative samples
+        to be 1 and the weight of positive samples to be the number of negative
+        samples divided by the number of positive samples.
+        """
+        task_pos_weights = []
+        n_tasks = len(self.tasks)
+
+        for i in range(n_tasks):
+            task_label = self.labels[:, i]
+            task_w = self.w[:, i]
+            task_label_ = task_label[task_w != 0]
+            num_positives = np.count_nonzero(task_label_)
+            num_negatives = len(task_label_) - num_positives
+
+            if num_positives > 0:
+                pos_weight = float(num_negatives) / num_positives
+                pos_indices = np.logical_and(task_label == 1, task_w != 0)
+                self.w[pos_indices, i] = pos_weight
+                task_pos_weights.append(pos_weight)
+            else:
+                task_pos_weights.append(1)
+
+        self.task_pos_weights = task_pos_weights
+
+    def split_dataset(self, frac_train, frac_val, frac_test):
+        """Split the dataset into three consecutive subsets for training,
+        validation and test.
+
+        Parameters
+        ----------
+        frac_train : float
+            Proportion of data to use for training
+        frac_val : float
+            Proportion of data to use for validation
+        frac_test : float
+            Proportion of data to use for test
+        """
+        frac_total = frac_train + frac_val + frac_test
+        assert frac_total == 1, \
```

Comment:
adding floats may have precision problem. Also I think we might need a BaseObject for all molecule dataset since there're multiple datasets for similar tasks.

Reply:
This is now replaced with `np.allclose(frac_total, 1.)`.